include/mesh/replicated_mesh.h

```diff
@@ -228,6 +228,13 @@ class ReplicatedMesh : public UnstructuredMesh
                         bool use_binary_search=true,
                         bool enforce_all_nodes_match_on_boundaries=false);
 
+  /**
+   * Return all points on boundary
+   * Note: The key of the returned map is a representative interior point of all disconnected subdomains.
+   *       The size of vector value is the number of disconnected boundaries for a subdomain.
+   */
+  std::map<Point, std::vector<std::vector<Point>>> get_boundary_points() const;
```

Comment:
This might be too hard to do soon on a distributed mesh, but should we move it to UnstructuredMesh and just rely on a MeshSerializer?  That might be good for flexibility when users want to triangulate a coarse mesh but expect to do a lot of refinement afterwards.

Reply:
Does MeshSerializer sends the mesh parts on other processors to the master so that master can have the full mesh? I was creating this function for mesh generation, the stage where mesh has not been distributed, refined, etc. I used some low-level functions in this function and my understanding on how libMesh sets up the local connectivity for various types of cells may not be accurate. I also have not thought of 3D carefully. Is the function written in a way easily understandable?

Reply:
The second optional argument to MeshSerializer lets you serialize to only rank 0, so you can run the algorithm on one processor with all the mesh parts sent there but without having to send them to other processors too.  If the mesh is currently serial, like when it hasn't been distributed yet, the function just tests that and then does nothing.  The destructor re-distributes the mesh (if it hadn't been serial to begin with) so the state change isn't permanent.  Basically you construct it at the beginning of whatever scope you need to be serial and then you're done.

Reply:
Is Point reliable with the less than operator? I don't think so. I've definitely run into issues using Points as keys in maps in the past. I would recommend against using a map there. Perhaps would should implement a hash function and use unordered_map instead?

Reply:
Point inherits TypeVector::operator<, which does the standard dictionary ordering.  And std::map accepts any total ordering, so Point should be fine as a key there.  Any more detailed recollection on those issues?  The only trouble I ever remember having was in the context of Point values changing: with coarsening and re-refinement I don't think we can guarantee that a fine node will be recreated in exactly the same place, due to floating point error.  That epsilon error would propagate to the centroids used here.  But I don't get the impression that in this context it's important for the choice or exact location of the representative points to be identical from run to run?

Reply:
This would have been 8+ years ago. The problem was that I was using it to lookup values where the Point precision might change by something close to machine precision (perhaps because of the way I was generating points using trigonometric functions). I seem to recall that if the x-value (the first test) was off by machine precision sometimes a large change in y or z wouldn't trigger the expected result because I would have taken the "wrong" branch on the x-coordinate first. I don't think this is a problem with Point per se, It's more of an issue with floating point equality, which you have to be cognizant of when using it as a map key.

Reply:
Oh, yes, the ordering here is definitely infinitely sensitive to Point coordinates in cases where two points are aligned on the yz plane.

Reply:
I see. Looks like MeshSerializer is a really powerful tool.

Using std::map directly creates a mapping for a representative point of a subdomain to its boundaries. We do not have to use map though.

Reply:
IIRC there's no good way around that, though, because any ordering here is essentially a function from 1D to fill 2D, and the inverse of such a function can't be continuous.

Reply:
I might suggest adding a hash function to Point (or TypeVector) so that it is more amenable to unordered_map usage (perhaps there's already one there, or the default hashing is fine). We have been running benchmarks here in the office and have discovered that `unordered_map` is almost always faster than `map` for both heavy insert and lookup cases. We are starting to lean on unordered_map more often when ordering isn't required (it often isn't).

Reply:
Good to know. Here the points should be well separated because each point is within one subdomain of a set of disconnected subdomains.

Reply:
Yes, using a representative element pointer `Elem *`  instead of `Point` could be a better idea.

Reply:
Oh, I'm definitely on board with using unordered_map everywhere you don't need the ordering.  Just pointing out that if we *do* need a consistent ordering, there's no way to get it from point locations that isn't sensitive to perturbations.  I suspect an unordered_map based on lowest element ID is the way to go here.

Reply:
> Here the points should be well separated because each point is within one subdomain of a set of disconnected subdomains.

Well, the issue here is that well separated in space doesn't imply well separated in ordering.  If you have elements that have the exact same X coordinate, then just rotating your mesh by epsilon degrees can completely switch around the ordering of those elements' centroids.  I think using Elem id (lowest id, if you really want maximum reproducability) is a safer way to go, given that you're already relying on it later and so anybody who wants reproduceable orderings from this is going to be disabling id renumbering already.

Reply:
> Is Point reliable with the less than operator? I don't think so.

It's definitely not reliable to use `set<Point>`, `map<Point, X>` etc! If there was a way to prevent these from even compiling, I would be in favor of using it...